File: Topcoder/Sherlock.java
// BEGIN CUT HERE
// PROBLEM STATEMENT
// 
Computer, find "BENADRYL CUCUMBERPATCH".
Did you mean the actor who played Sherlock, Benedict Cumberbatch?


Benedict Cumberbatch is an actor who is famous not only because of his roles.
It is also known that you can misspell his name almost arbitrarily and people (and computers too!) will still recognize him.


In this problem, you are given the name of a person: the Strings firstName and lastName.
We say that this name matches Benedict Cumberbatch if it has all properties listed below.


Each name must have at least seven letters.
The first name must start with 'B' and the last name must start with 'C'.
At least three characters of firstName must appear somewhere in the name "BENEDICT".
At least five characters of lastName must appear somewhere in the name "CUMBERBATCH".


In conditions 3 and 4 we count each occurrence of a letter, even if that letter occurs in the given name more times than in the actor's actual name.
For example, "BATAXXAT" is a valid first name because three of its characters ('B', 'T', 'T') appear in "BENEDICT".


Return "It is him" if all conditions are satisfied and "It is someone else" if they aren't.


DEFINITION
Class:Sherlock
Method:isItHim
Parameters:String, String
Returns:String
Method signature:String isItHim(String firstName, String lastName)


NOTES
-Note that the return value is case-sensitive.


CONSTRAINTS
-firstName and lastName will each consist of between 1 and 20 characters, inclusive.
-Each character in firstName and lastName will be an uppercase English letter ('A'-'Z').


EXAMPLES

0)
"BENEDICT"
"CUMBERBATCH"

Returns: "It is him"



1)
"BENADRYL"
"CUCUMBERPATCH"

Returns: "It is him"



2)
"HARSHIT"
"MEHTA"

Returns: "It is someone else"



3)
"BATAXXAT"
"CURMUDGEON"

Returns: "It is him"



4)
"BENEDI"
"CUMBER"

Returns: "It is someone else"

Both names are too short.

5)
"BAXAXAXY"
"CENTRIFUGAL"

Returns: "It is someone else"

The first name has too few characters in common with BENEDICT.

6)
"BUMBLESHACK"
"CRUMPLEHORN"

Returns: "It is him"



// END CUT HERE
import java.util.*;
public class Sherlock {
	public String isItHim(String firstName, String lastName) {
		
	}
	public static void main(String[] args) {
		Sherlock temp = new Sherlock();
		System.out.println(temp.isItHim(String firstName, String lastName));
	}
}

File: Topcoder/Sherlock.cpp
#pragma GCC optimize(3)
#include<bits/stdc++.h>
#define MAXN 100005
#define INF 1000000000
#define MOD 1000000007
#define F first
#define S second
using namespace std;
typedef long long ll;
typedef pair<int,int> P;
class Sherlock {
	public:
	string isItHim(string firstName, string lastName) {
		
	}
};
// BEGIN CUT HERE
#include <cstdio>
#include <ctime>
#include <iostream>
#include <string>
#include <vector>
namespace moj_harness {
	using std::string;
	using std::vector;
	int run_test_case(int);
	void run_test(int casenum = -1, bool quiet = false) {
		if (casenum != -1) {
			if (run_test_case(casenum) == -1 && !quiet) {
				std::cerr << "Illegal input! Test case " << casenum << " does not exist." << std::endl;
			}
			return;
		}
		
		int correct = 0, total = 0;
		for (int i=0;; ++i) {
			int x = run_test_case(i);
			if (x == -1) {
				if (i >= 100) break;
				continue;
			}
			correct += x;
			++total;
		}
		
		if (total == 0) {
			std::cerr << "No test cases run." << std::endl;
		} else if (correct < total) {
			std::cerr << "Some cases FAILED (passed " << correct << " of " << total << ")." << std::endl;
		} else {
			std::cerr << "All " << total << " tests passed!" << std::endl;
		}
	}
	
	int verify_case(int casenum, const string &expected, const string &received, std::clock_t elapsed) { 
		std::cerr << "Example " << casenum << "... "; 
		
		string verdict;
		vector<string> info;
		char buf[100];
		
		if (elapsed > CLOCKS_PER_SEC / 200) {
			std::sprintf(buf, "time %.2fs", elapsed * (1.0/CLOCKS_PER_SEC));
			info.push_back(buf);
		}
		
		if (expected == received) {
			verdict = "PASSED";
		} else {
			verdict = "FAILED";
		}
		
		std::cerr << verdict;
		if (!info.empty()) {
			std::cerr << " (";
			for (size_t i=0; i<info.size(); ++i) {
				if (i > 0) std::cerr << ", ";
				std::cerr << info[i];
			}
			std::cerr << ")";
		}
		std::cerr << std::endl;
		
		if (verdict == "FAILED") {
			std::cerr << "    Expected: \"" << expected << "\"" << std::endl; 
			std::cerr << "    Received: \"" << received << "\"" << std::endl; 
		}
		
		return verdict == "PASSED";
	}

	int run_test_case(int casenum__) {
		switch (casenum__) {
		case 0: {
			string firstName          = "BENEDICT";
			string lastName           = "CUMBERBATCH";
			string expected__         = "It is him";

			std::clock_t start__      = std::clock();
			string received__         = Sherlock().isItHim(firstName, lastName);
			return verify_case(casenum__, expected__, received__, clock()-start__);
		}
		case 1: {
			string firstName          = "BENADRYL";
			string lastName           = "CUCUMBERPATCH";
			string expected__         = "It is him";

			std::clock_t start__      = std::clock();
			string received__         = Sherlock().isItHim(firstName, lastName);
			return verify_case(casenum__, expected__, received__, clock()-start__);
		}
		case 2: {
			string firstName          = "HARSHIT";
			string lastName           = "MEHTA";
			string expected__         = "It is someone else";

			std::clock_t start__      = std::clock();
			string received__         = Sherlock().isItHim(firstName, lastName);
			return verify_case(casenum__, expected__, received__, clock()-start__);
		}
		case 3: {
			string firstName          = "BATAXXAT";
			string lastName           = "CURMUDGEON";
			string expected__         = "It is him";

			std::clock_t start__      = std::clock();
			string received__         = Sherlock().isItHim(firstName, lastName);
			return verify_case(casenum__, expected__, received__, clock()-start__);
		}
		case 4: {
			string firstName          = "BENEDI";
			string lastName           = "CUMBER";
			string expected__         = "It is someone else";

			std::clock_t start__      = std::clock();
			string received__         = Sherlock().isItHim(firstName, lastName);
			return verify_case(casenum__, expected__, received__, clock()-start__);
		}
		case 5: {
			string firstName          = "BAXAXAXY";
			string lastName           = "CENTRIFUGAL";
			string expected__         = "It is someone else";

			std::clock_t start__      = std::clock();
			string received__         = Sherlock().isItHim(firstName, lastName);
			return verify_case(casenum__, expected__, received__, clock()-start__);
		}
		case 6: {
			string firstName          = "BUMBLESHACK";
			string lastName           = "CRUMPLEHORN";
			string expected__         = "It is him";

			std::clock_t start__      = std::clock();
			string received__         = Sherlock().isItHim(firstName, lastName);
			return verify_case(casenum__, expected__, received__, clock()-start__);
		}

		// custom cases

/*      case 7: {
			string firstName          = ;
			string lastName           = ;
			string expected__         = ;

			std::clock_t start__      = std::clock();
			string received__         = Sherlock().isItHim(firstName, lastName);
			return verify_case(casenum__, expected__, received__, clock()-start__);
		}*/
/*      case 8: {
			string firstName          = ;
			string lastName           = ;
			string expected__         = ;

			std::clock_t start__      = std::clock();
			string received__         = Sherlock().isItHim(firstName, lastName);
			return verify_case(casenum__, expected__, received__, clock()-start__);
		}*/
/*      case 9: {
			string firstName          = ;
			string lastName           = ;
			string expected__         = ;

			std::clock_t start__      = std::clock();
			string received__         = Sherlock().isItHim(firstName, lastName);
			return verify_case(casenum__, expected__, received__, clock()-start__);
		}*/
		default:
			return -1;
		}
	}
}

#include <cstdlib>
int main(int argc, char *argv[]) {
	if (argc == 1) {
		moj_harness::run_test();
	} else {
		for (int i=1; i<argc; ++i)
			moj_harness::run_test(std::atoi(argv[i]));
	}
}
// END CUT HERE
